Sources/SwiftFusion/Core/Vector.swift.gyb

```diff
@@ -0,0 +1,84 @@
+import TensorFlow
+
+% dims = [1, 2, 3]
+% for dim in dims:
+% coordinates = ['x', 'y', 'z'][0:dim]
+
+/// An element of R^${dim}.
+public struct Vector${dim}:
+  AdditiveArithmetic, Differentiable, ElementaryFunctions, KeyPathIterable, TangentStandardBasis
```

Comment:
I see what you are trying to achieve here, but can we have something that is directly compatible with a TF Tensor?

Reply:
This is as close as we can get with a wrapper type -- you can call `.tensor` to get the TF Tensor and you can do `VectorN(t)` to convert a TF Tensor to one of these.

We could use a simple unwrapped TF Tensor. I like the static type information about the dimension that we get from the wrapper, but I wouldn't object to just using TF Tensor if you think that would be better.

Reply:
I like the small vector types - we also have this in GTSAM with static Eigen vectors. I think it is key to performance, and the static type checking is great. It does add extra code and complexity...

Reply:
I had a somewhat related thought when thinking about this.

We could allow a "batch dimension" in these vectors so that for example a `Vector2` is actually a `2 x batchSize`. Then, if we write all operations to distribute over the batch dimension correctly, we can pack all the vectors that we're operating on together and do the operations on all the vectors at the same time. This may speed things up a lot.

What do you think? Does GTSAM do something similar anywhere?

Reply:
Yea I agree on that :)

Reply:
KISS :-) feels like premature optimization at this point?

Reply:
I’m totally open to these type of things but I’d focus on the mission now. Which is to get Gauss Newton and CGLS going on PoseSLAM :-) we can then measure...